docs/debugging.md

```diff
@@ -271,4 +271,90 @@ slower than a compiled version, so you should default to installing the wheel in
 CI instead of compiling Cython, which can take up to a few minutes on some CI
 runners.
 
+### Compiling CPython and foundational packages with thread sanitizer (TSAN)
+
+[Thread sanitizer](https://github.com/google/sanitizers/wiki/ThreadSanitizerCppManual) (or TSAN) helps
+to detect C/C++ data races in concurrent systems. This tool can help to reveal free-threading
+related bugs in CPython and foundational packages (e.g. `numpy`).
+In this section we provide the commands to build free-threading compatible CPython and packages with TSAN and other hints to discover potential data races.
+
+#### Compile free-threading CPython with TSAN
+
+- Clone free-threading stable branch, e.g. 3.13.1
+
+```bash
+git clone https://github.com/python/cpython.git -b v3.13.1
+```
+
+- Configure and build (we use clang-18 as compiler):
+
+```bash
+cd cpython
+CC=clang-18 CXX=clang++-18 ./configure --disable-gil --with-thread-sanitizer --prefix $PWD/cpython-tsan
+make -j 8
+make install
+```
+
+- Use built python interpreter:
+
+```bash
+export PATH=$PWD/cpython-tsan/bin:$PATH
```

Comment:
I personally find it nice to work in a venv at this point, so:

```
$PWD/cpython-tsan/bin/python3.13t -m venv ~/tsanvenv
source ~/tsanvenv/bin/activate
```

and once you've done that you can just say `python` instead of `python3.13t` for the rest of the document.